Bootloader/firmware/src/app.c

```diff
@@ -35,20 +62,45 @@ typedef struct {
   USB_DEVICE_HANDLE usb_device;  //!< The USB Device layer handle.
   AppState state;
   DFUState dfu_state;
-  DFUStatus dfu_status;
-  uint8_t alternate_setting;
+  DFUStatus dfu_status;  //!< The current DFU state.
+  void *write_address;  //!< The address to write the chunk of data to.
+  uint16_t next_block;  //!< The expected index of the next block received.
+  uint16_t data_length;  //!< The length of the data in DATA_BUFFER.
   bool is_configured;  //!< Keep track of whether the device is configured.
-  bool has_new_firmware;
+  bool erase_flash;  //!< True if we should start a flash erase cycle
+  bool has_new_firmware;  //!< True if there is new firmware ready.
 } AppData;
 
-static uint8_t STATUS_RESPONSE[GET_STATUS_RESPONSE_SIZE];
 static AppData g_app;
-static uint8_t DATA_BUFFER[DFU_BLOCK_SIZE];
+
+/**
+ * @brief The buffer that holds the DFU Status response.
+ */
+static uint8_t g_status_response[GET_STATUS_RESPONSE_SIZE];
+
+/*
+ * @brief The buffer into which we receive DFU data.
+ *
+ * The minimum flash program size is a word (32-bits). I can't see anything in
+ * the DFU standard that requires the transfer size to be a multiple of 4. This
+ * means we could end up with 1-3 bytes that we can't write immediately.
+ *
+ * The solution is to make the incoming DFU buffer slightly bigger so we can
+ * accomodate this, and write them out once we receive the remaining data.
+ */
+static uint8_t DATA_BUFFER[DFU_BLOCK_SIZE + FLASH_WORD_SIZE - 1];
 
 // Helper functions
 // ----------------------------------------------------------------------------
+static inline uint32_t ExtractUInt32(const uint8_t *ptr) {
```

Comment:
There is no utils. Almost all the bootloader code is in a single file. If I need to use it elsewhere I'll move it out.